plugins/inputs/modbus/request.go

```diff
@@ -9,24 +12,169 @@ type request struct {
+		}
+		start = end
+
+		if len(current.fields) > 0 {
+			requests = append(requests, current)
+		}
+	}
+
+	return requests
+}
+
+func shrinkGroup(g request, maxBatchSize uint16) []request {
+	var requests []request
+	var current request
+
+	for _, f := range g.fields {
+		// Just add the field and update length if we are still
+		// within the maximum batch-size
+		if current.length > 0 && f.address+f.length <= current.address+maxBatchSize {
+			current.fields = append(current.fields, f)
+			current.length = f.address - current.address + f.length
+			continue
+		}
+
+		// Ignore completely empty requests
+		if len(current.fields) > 0 {
+			requests = append(requests, current)
+		}
+
+		// Create a new request
+		current = request{
+			fields:  []field{f},
+			address: f.address,
+			length:  f.length,
+		}
+	}
+	if len(current.fields) > 0 {
+		requests = append(requests, current)
+	}
+
+	return requests
+}
+
+func optimitzeGroup(g request, maxBatchSize uint16) []request {
+	if len(g.fields) == 0 {
+		return nil
+	}
+
+	requests := shrinkGroup(g, maxBatchSize)
+	length := countRegisters(requests)
+
+	for i := 1; i < len(g.fields)-1; i++ {
+		// Always keep consecutive fields as they are known to be optimal
+		if g.fields[i-1].address+g.fields[i-1].length == g.fields[i].address {
+			continue
+		}
+
+		// Perform the split and check if it is better
+		// Note: This involves recursive optimization of the right side of the split.
+		current := shrinkGroup(request{fields: g.fields[:i]}, maxBatchSize)
+		current = append(current, optimitzeGroup(request{fields: g.fields[i:]}, maxBatchSize)...)
+		currentLength := countRegisters(current)
+
+		// Do not allow for more requests
+		if len(current) > len(requests) {
+			continue
+		}
+		// Try to reduce the number of registers we are trying to access
+		if currentLength >= length {
+			continue
+		}
+
+		// We found a better solution
+		requests = current
+		length = currentLength
+	}
+
+	return requests
+}
+
+func optimitzeGroupWithinLimits(fields []field, maxBatchSize uint16, maxExtraRegisters uint16) []request {
+	fields = removeAllOmitted(fields)
+	var requests []request
+	currentRequest := newRequest(fields[0])
+	for _, f := range fields[1:] {
+		// Check if we need to interrupt the current chunk and require a new one
+		extraLength := f.address + f.length - (currentRequest.address + currentRequest.length)
+		needInterrupt := extraLength > maxExtraRegisters                                  // too far appart
+		needInterrupt = needInterrupt || currentRequest.length+extraLength > maxBatchSize // too large
+		if !needInterrupt {
+			// Still safe to add the field to the current request
+			currentRequest.length = f.address + f.length - currentRequest.address
+			currentRequest.fields = append(currentRequest.fields, f)
+			continue
+		}
+		// Finish the current request, add it to the list and construct a new one
+		requests = append(requests, currentRequest)
+		currentRequest = newRequest(f)
+
+	}
```

Comment:
@TimurDela this is the only remaining CI issue I believe, you have an extra line here

```suggestion
		currentRequest = newRequest(f)
	}
```

Reply:
Thanks. Which part of the CI error message told you that this was the problem?

Reply:
For linting issues, CI will leave a comment on your PR with the issue.